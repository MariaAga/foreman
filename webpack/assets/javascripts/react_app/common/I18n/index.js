import intl from './intl.js';
import { jed, translate, ngettext, sprintf } from './jed.js';

export const documentLocale = async () => {
  await new Promise(resolve => setImmediate(resolve));
  return intl.locale;
};

export { translate, ngettext, jed, sprintf, intl };

export default { translate, ngettext, jed, sprintf, intl };

window.__ = translate;
window.n__ = ngettext;
