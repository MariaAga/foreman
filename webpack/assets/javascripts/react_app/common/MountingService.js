import ReactDOM from 'react-dom';
import React, { useState, useEffect } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';

export { default as registerReducer } from '../redux/reducers/registerReducer';

// eslint-disable-next-line react/prop-types
const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const [allPluginsImported, setAllPluginsImported] = useState(false);
  async function mountComponent() {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else if (allPluginsImported) {
      const awaitedComponent = componentRegistry.markup(component, {
        data,
        store,
        flattenData,
      });
      setMounted(true);
      setMountedComponent(awaitedComponent);
    }
  }
  const updateAllPluginsImported = e => {
    if (Object.values(window.allPluginsLoaded).every(Boolean))
      setAllPluginsImported(true);
  };
  useEffect(() => {
    document.addEventListener('loadPlugin', updateAllPluginsImported);
    return () =>
      window.removeEventListener('loadPlugin', updateAllPluginsImported);
  }, []);
  useEffect(() => {
    if (!mounted) mountComponent();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [allPluginsImported]);
  return mounted ? mountedComponent : <div>Loading...</div>;
};

function mountNode(component, reactNode, data, flattenData) {
  ReactDOM.render(
    <AwaitedMount
      component={component}
      data={data}
      flattenData={flattenData}
    />,
    reactNode
  );
}

/**
 * This is a html tag (Web component) that can be used for mounting react component from ComponentRegistry.
 */
class ReactComponentElement extends HTMLElement {
  static get observedAttributes() {
    return ['data-props'];
  }

  get componentName() {
    return this.getAttribute('name');
  }
  get reactProps() {
    return this.dataset.props !== '' ? JSON.parse(this.dataset.props) : {};
  }
  set reactProps(newProps) {
    this.dataset.props = JSON.stringify(newProps);
  }
  get mountPoint() {
    if (!this._mountPoint) {
      this._mountPoint = this;
    }

    return this._mountPoint;
  }

  attributeChangedCallback(name, oldValue, newValue) {
    switch (name) {
      case 'data-props':
        // if this is not the initial prop set
        if (oldValue !== null) this._render();
        break;
      default:
      // We don't know how to react to default attribute change
    }
  }

  connectedCallback() {
    this._render();
  }

  _render() {
    try {
      mountNode(this.componentName, this, this.reactProps, true);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(
        `Unable to mount foreman-react-component: ${this.componentName}`,
        error
      );
    }
  }
}

if (!window.customElements.get('foreman-react-component')) {
  window.customElements.define(
    'foreman-react-component',
    ReactComponentElement
  );
}
