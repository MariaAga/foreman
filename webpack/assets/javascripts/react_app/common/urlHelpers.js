/**
 * Build a url from given controller, action and id
 * @param {String} controller - the controller
 * @param {String} action - the action
 */
export const urlBuilder = (controller, action, id = undefined) =>
  `/${controller}/${id ? `${id}/` : ''}${action}`;

/**
 * Build a url with search query
 * @param {String} base - the base url
 * @param {String} searchQuery - the search query
 */
export const urlWithSearch = (base, searchQuery) =>
  `/${base}?search=${searchQuery}`;
