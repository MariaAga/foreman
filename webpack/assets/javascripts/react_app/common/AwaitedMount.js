import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';
import { translate as __ } from './I18n';

export const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const [allPluginsImported, setAllPluginsImported] = useState(false);
  async function mountComponent() {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else if (allPluginsImported) {
      const awaitedComponent = componentRegistry.markup(component, {
        data,
        store,
        flattenData,
      });
      setMounted(true);
      setMountedComponent(awaitedComponent);
    }
  }
  const updateAllPluginsImported = e => {
    setAllPluginsImported(true);
  };
  useEffect(() => {
    document.addEventListener('loadJS', updateAllPluginsImported);
    return () => window.removeEventListener('loadJS', updateAllPluginsImported);
  }, []);
  useEffect(() => {
    if (!mounted) mountComponent();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [allPluginsImported]);
  return mounted ? mountedComponent : <div>{__('Loading...')}</div>;
};

AwaitedMount.propTypes = {
  component: PropTypes.string.isRequired,
  data: PropTypes.object,
  flattenData: PropTypes.bool,
};
AwaitedMount.defaultProps = {
  data: {},
  flattenData: false,
};
