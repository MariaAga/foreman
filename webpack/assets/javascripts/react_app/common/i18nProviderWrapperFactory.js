import React from 'react';
import { IntlProvider } from 'react-intl';
import { intl } from './I18n';
import { getDisplayName } from './helpers';

const i18nProviderWrapperFactory = () => WrappedComponent => {
  const wrappedName = getDisplayName(WrappedComponent);
  const I18nProviderWrapper = (props) => {
    const [htmlElemnt] = document.getElementsByTagName('html');
    const langAttr = htmlElemnt.getAttribute('lang') || intl.locale;
    const timezone = htmlElemnt.getAttribute('data-timezone') || intl.timezone;
    return (
      <IntlProvider
        locale={langAttr}
        timeZone={timezone}
      >
          <WrappedComponent {...props} />
      </IntlProvider>
    );
  };

  I18nProviderWrapper.displayName = `I18nProviderWrapper(${wrappedName})`;

  return I18nProviderWrapper;
};

export { i18nProviderWrapperFactory };
