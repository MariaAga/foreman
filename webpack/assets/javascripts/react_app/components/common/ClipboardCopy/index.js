import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button } from 'patternfly-react';
import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import { translate as __ } from '../../../common/I18n';
import './clipboard-copy.scss';

const ClipboardCopy = ({
  text: defaultText,
  successMessage,
  buttonText,
  textareaProps,
  buttonProps,
}) => {
  const [text, setText] = useState(defaultText);

  return (
    <div className="clipboard-copy">
      <textarea
        defaultValue={text}
        onChange={({ target: { value } }) => setText(value)}
        {...textareaProps}
      />
      <Tooltip
        content={successMessage}
        position={TooltipPosition.right}
        trigger="click"
      >
        <Button
          onClick={() => navigator.clipboard.writeText(text)}
          bsStyle="default"
          {...buttonProps}
        >
          {buttonText}
        </Button>
      </Tooltip>
    </div>
  );
};

ClipboardCopy.propTypes = {
  text: PropTypes.string.isRequired,
  buttonText: PropTypes.string,
  successMessage: PropTypes.string,
  textareaProps: PropTypes.object,
  buttonProps: PropTypes.object,
};

ClipboardCopy.defaultProps = {
  buttonText: __('Copy to clipboard'),
  successMessage: __('Copied!'),
  textareaProps: {},
  buttonProps: {},
};

export default ClipboardCopy;
