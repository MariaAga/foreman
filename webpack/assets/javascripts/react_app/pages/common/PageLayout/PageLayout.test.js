import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';
import PageLayout from './PageLayout';
import { pageLayoutMock } from './PageLayout.fixtures';

const pageLayoutFixtures = {
  'render pageLayout w/search': pageLayoutMock,
  'render pageLayout without search': { ...pageLayoutMock, searchable: false },
  'render pageLayout with custom breadcrumbs': {
    ...pageLayoutMock,
    customBreadcrumbs: 'customBreadcrumbs',
  },
  'render pageLayout without breadcrumbs': {
    ...pageLayoutMock,
    breadcrumbOptions: null,
  },
  'render pageLayout w/toastNotifications': {
    ...pageLayoutMock,
    toastNotifications: 'notification',
  },
  'render pageLayout w/toolBar': {
    ...pageLayoutMock,
    toolbarButtons: 'toolbarButton',
  },
  'render pageLayout w/Export': {
    ...pageLayoutMock,
    exportURL: 'urlExport',
  },
  'render pageLayout w/Docs': {
    ...pageLayoutMock,
    documentationURL: 'documentationURL',
  },
};

testComponentSnapshotsWithFixtures(PageLayout, pageLayoutFixtures);
