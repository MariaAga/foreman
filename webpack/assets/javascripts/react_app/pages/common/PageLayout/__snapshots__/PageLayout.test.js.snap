// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render pageLayout w/Docs 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      id="breadcrumb"
    >
      <Connect(BreadcrumbBar)
        data={
          Object {
            "data": Object {
              "breadcrumbItems": Array [
                Object {
                  "caption": "root",
                  "url": "/some-url",
                },
                Object {
                  "caption": "child with onClick",
                  "onClick": [MockFunction],
                },
                Object {
                  "caption": "active child",
                },
              ],
              "isSwitchable": false,
              "resource": Object {
                "nameField": "name",
                "resourceUrl": "some/url",
                "switcherItemUrl": "some/url/:id",
              },
            },
          }
        }
      />
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={6}
      >
        <div
          id="search-bar"
        >
          <Connect(SearchBar)
            data={
              Object {
                "data": Object {
                  "autocomplete": Object {
                    "id": "some-id",
                    "searchQuery": null,
                    "url": "model/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "bookmarks": Object {
                    "canCreate": true,
                    "documentationUrl": "/doc/url",
                    "url": "/api/bookmarks",
                  },
                  "controller": "models",
                },
              }
            }
            initialQuery=""
            onBookmarkClick={[Function]}
            onSearch={[Function]}
          />
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={6}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
          <Connect(Fill)
            id="documentationButton"
            key="documentationButton"
            slotId="layout-toolbar-button"
            weight={200}
          >
            <DocumentationButton
              text="Documentation"
              url="documentationURL"
            />
          </Connect(Fill)>
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;

exports[`render pageLayout w/Export 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      id="breadcrumb"
    >
      <Connect(BreadcrumbBar)
        data={
          Object {
            "data": Object {
              "breadcrumbItems": Array [
                Object {
                  "caption": "root",
                  "url": "/some-url",
                },
                Object {
                  "caption": "child with onClick",
                  "onClick": [MockFunction],
                },
                Object {
                  "caption": "active child",
                },
              ],
              "isSwitchable": false,
              "resource": Object {
                "nameField": "name",
                "resourceUrl": "some/url",
                "switcherItemUrl": "some/url/:id",
              },
            },
          }
        }
      />
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={6}
      >
        <div
          id="search-bar"
        >
          <Connect(SearchBar)
            data={
              Object {
                "data": Object {
                  "autocomplete": Object {
                    "id": "some-id",
                    "searchQuery": null,
                    "url": "model/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "bookmarks": Object {
                    "canCreate": true,
                    "documentationUrl": "/doc/url",
                    "url": "/api/bookmarks",
                  },
                  "controller": "models",
                },
              }
            }
            initialQuery=""
            onBookmarkClick={[Function]}
            onSearch={[Function]}
          />
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={6}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
          <Connect(Fill)
            id="exportButton"
            key="exportButton"
            slotId="layout-toolbar-button"
            weight={100}
          >
            <ExportButton
              text="Export"
              title="Export to CSV"
              url="urlExport"
            />
          </Connect(Fill)>
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;

exports[`render pageLayout w/search 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      id="breadcrumb"
    >
      <Connect(BreadcrumbBar)
        data={
          Object {
            "data": Object {
              "breadcrumbItems": Array [
                Object {
                  "caption": "root",
                  "url": "/some-url",
                },
                Object {
                  "caption": "child with onClick",
                  "onClick": [MockFunction],
                },
                Object {
                  "caption": "active child",
                },
              ],
              "isSwitchable": false,
              "resource": Object {
                "nameField": "name",
                "resourceUrl": "some/url",
                "switcherItemUrl": "some/url/:id",
              },
            },
          }
        }
      />
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={6}
      >
        <div
          id="search-bar"
        >
          <Connect(SearchBar)
            data={
              Object {
                "data": Object {
                  "autocomplete": Object {
                    "id": "some-id",
                    "searchQuery": null,
                    "url": "model/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "bookmarks": Object {
                    "canCreate": true,
                    "documentationUrl": "/doc/url",
                    "url": "/api/bookmarks",
                  },
                  "controller": "models",
                },
              }
            }
            initialQuery=""
            onBookmarkClick={[Function]}
            onSearch={[Function]}
          />
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={6}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;

exports[`render pageLayout w/toastNotifications 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      data-notifications="notification"
      id="toast-notifications-container"
    >
      <Connect(ToastsList) />
    </div>
    <div
      id="breadcrumb"
    >
      <Connect(BreadcrumbBar)
        data={
          Object {
            "data": Object {
              "breadcrumbItems": Array [
                Object {
                  "caption": "root",
                  "url": "/some-url",
                },
                Object {
                  "caption": "child with onClick",
                  "onClick": [MockFunction],
                },
                Object {
                  "caption": "active child",
                },
              ],
              "isSwitchable": false,
              "resource": Object {
                "nameField": "name",
                "resourceUrl": "some/url",
                "switcherItemUrl": "some/url/:id",
              },
            },
          }
        }
      />
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={6}
      >
        <div
          id="search-bar"
        >
          <Connect(SearchBar)
            data={
              Object {
                "data": Object {
                  "autocomplete": Object {
                    "id": "some-id",
                    "searchQuery": null,
                    "url": "model/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "bookmarks": Object {
                    "canCreate": true,
                    "documentationUrl": "/doc/url",
                    "url": "/api/bookmarks",
                  },
                  "controller": "models",
                },
              }
            }
            initialQuery=""
            onBookmarkClick={[Function]}
            onSearch={[Function]}
          />
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={6}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;

exports[`render pageLayout w/toolBar 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      id="breadcrumb"
    >
      <Connect(BreadcrumbBar)
        data={
          Object {
            "data": Object {
              "breadcrumbItems": Array [
                Object {
                  "caption": "root",
                  "url": "/some-url",
                },
                Object {
                  "caption": "child with onClick",
                  "onClick": [MockFunction],
                },
                Object {
                  "caption": "active child",
                },
              ],
              "isSwitchable": false,
              "resource": Object {
                "nameField": "name",
                "resourceUrl": "some/url",
                "switcherItemUrl": "some/url/:id",
              },
            },
          }
        }
      />
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={6}
      >
        <div
          id="search-bar"
        >
          <Connect(SearchBar)
            data={
              Object {
                "data": Object {
                  "autocomplete": Object {
                    "id": "some-id",
                    "searchQuery": null,
                    "url": "model/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "bookmarks": Object {
                    "canCreate": true,
                    "documentationUrl": "/doc/url",
                    "url": "/api/bookmarks",
                  },
                  "controller": "models",
                },
              }
            }
            initialQuery=""
            onBookmarkClick={[Function]}
            onSearch={[Function]}
          />
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={6}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;

exports[`render pageLayout with custom breadcrumbs 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      id="breadcrumb"
    >
      <Component />
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={6}
      >
        <div
          id="search-bar"
        >
          <Connect(SearchBar)
            data={
              Object {
                "data": Object {
                  "autocomplete": Object {
                    "id": "some-id",
                    "searchQuery": null,
                    "url": "model/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "bookmarks": Object {
                    "canCreate": true,
                    "documentationUrl": "/doc/url",
                    "url": "/api/bookmarks",
                  },
                  "controller": "models",
                },
              }
            }
            initialQuery=""
            onBookmarkClick={[Function]}
            onSearch={[Function]}
          />
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={6}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;

exports[`render pageLayout without breadcrumbs 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      id="breadcrumb"
    >
      <div
        className="row form-group"
      >
        <h1
          className="col-md-8"
        >
          Page
        </h1>
      </div>
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={6}
      >
        <div
          id="search-bar"
        >
          <Connect(SearchBar)
            data={
              Object {
                "data": Object {
                  "autocomplete": Object {
                    "id": "some-id",
                    "searchQuery": null,
                    "url": "model/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "bookmarks": Object {
                    "canCreate": true,
                    "documentationUrl": "/doc/url",
                    "url": "/api/bookmarks",
                  },
                  "controller": "models",
                },
              }
            }
            initialQuery=""
            onBookmarkClick={[Function]}
            onSearch={[Function]}
          />
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={6}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;

exports[`render pageLayout without search 1`] = `
<div
  id="main"
>
  <div
    id="react-content"
  >
    <div
      id="breadcrumb"
    >
      <Connect(BreadcrumbBar)
        data={
          Object {
            "data": Object {
              "breadcrumbItems": Array [
                Object {
                  "caption": "root",
                  "url": "/some-url",
                },
                Object {
                  "caption": "child with onClick",
                  "onClick": [MockFunction],
                },
                Object {
                  "caption": "active child",
                },
              ],
              "isSwitchable": false,
              "resource": Object {
                "nameField": "name",
                "resourceUrl": "some/url",
                "switcherItemUrl": "some/url/:id",
              },
            },
          }
        }
      />
    </div>
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        className="title_filter"
        componentClass="div"
        md={4}
      />
      <Col
        bsClass="col"
        componentClass="div"
        id="title_action"
        md={8}
      >
        <div
          className="btn-toolbar pull-right"
        >
          <Connect(Slot)
            id="layout-toolbar-button"
            multi={true}
          />
        </div>
      </Col>
    </Row>
    body
  </div>
</div>
`;
